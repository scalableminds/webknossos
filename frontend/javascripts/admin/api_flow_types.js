/**
 * api_flow_types.js
 * @flow
 */
import Enum from "Enumjs";

import type {
  BoundingBoxObject,
  Edge,
  CommentType,
  TreeGroup,
  RecommendedConfiguration,
} from "oxalis/store";
import type { ServerUpdateAction } from "oxalis/model/sagas/update_actions";
import type { SkeletonTracingStats } from "oxalis/model/accessors/skeletontracing_accessor";
import type { Vector3, Vector6, Point3 } from "oxalis/constants";

export type APIMessage = { ["info" | "warning" | "error"]: string };

export type ElementClass =
  | "uint8"
  | "uint16"
  | "uint24"
  | "uint32"
  | "uint64"
  | "float"
  | "double"
  | "int8"
  | "int16"
  | "int32"
  | "int64";

export type APIMapping = {
  +parent?: string,
  +name: string,
  +classes?: Array<Array<number>>,
  +colors?: Array<number>,
  +hideUnmappedIds?: boolean,
};

type APIDataLayerBase = {|
  +name: string,
  +boundingBox: BoundingBoxObject,
  +resolutions: Array<Vector3>,
  +elementClass: ElementClass,
|};

type APIColorLayer = {|
  ...APIDataLayerBase,
  +category: "color",
|};

export type APISegmentationLayer = {|
  ...APIDataLayerBase,
  +category: "segmentation",
  +largestSegmentId: number,
  +mappings?: Array<string>,
  +fallbackLayer?: ?string,
|};

export type APIDataLayer = APIColorLayer | APISegmentationLayer;

export type APIHistogramData = Array<{
  numberOfElements: number,
  elementCounts: Array<number>,
  min: number,
  max: number,
}>;

type APIDataSourceBase = {
  +id: {
    +name: string,
    +team: string,
  },
  +status?: string,
};

export type APIMaybeUnimportedDataSource = APIDataSourceBase & {
  +dataLayers?: Array<APIDataLayer>,
  +scale: ?Vector3,
};

export type APIDataSource = APIDataSourceBase & {
  +dataLayers: Array<APIDataLayer>,
  +scale: Vector3,
};

export type APIDataStore = {
  +name: string,
  +url: string,
  +isForeign: boolean,
  +isScratch: boolean,
  +isConnector: boolean,
};

export type APITracingStore = {
  +name: string,
  +url: string,
};

export type APITeam = {
  +id: string,
  +name: string,
  +organization: string,
};

type APIPublication = {
  +created: number,
  +description: string,
  +id: string,
  +imageUrl: string,
  +publicationDate: number,
  +title: string,
};

export type APIDatasetId = {
  +owningOrganization: string,
  +name: string,
};

export type APIDatasetDetails = {
  +species?: string,
  +brainRegion?: string,
  +acquisition?: string,
};

type APIDatasetBase = APIDatasetId & {
  +isUnreported: boolean,
  +allowedTeams: Array<APITeam>,
  +created: number,
  +dataStore: APIDataStore,
  +description: ?string,
  +details: ?APIDatasetDetails,
  +isEditable: boolean,
  +isPublic: boolean,
  +displayName: ?string,
  +logoUrl: ?string,
  +lastUsedByUser: number,
  +isForeign: boolean,
  +sortingKey: number,
  +publication: ?APIPublication,
};

export type APIMaybeUnimportedDataset = APIDatasetBase & {
  +dataSource: APIMaybeUnimportedDataSource,
  +isActive: boolean,
};

export type APIDataset = APIDatasetBase & {
  +dataSource: APIDataSource,
  +isActive: true,
};

export type APISampleDataset = {
  +name: string,
  +description: string,
  +status: "available" | "downloading" | "present",
};

export type APIDataSourceWithMessages = {
  +dataSource?: APIDataSource,
  +previousDataSource?: APIDataSource,
  +messages: Array<APIMessage>,
};

export type APITeamMembership = {
  +id: string,
  +name: string,
  +isTeamManager: boolean,
};

export type ExperienceMap = { +[string]: number };

export type ExperienceDomainList = Array<string>;

export type APIUserBase = {
  +email: string,
  +firstName: string,
  +lastName: string,
  +id: string,
  +isAnonymous: boolean,
  +teams: Array<APITeamMembership>,
};

export type APIUser = APIUserBase & {
  +created: number,
  +experiences: ExperienceMap,
  +isAdmin: boolean,
  +isActive: boolean,
  +isEditable: boolean,
  +lastActivity: number,
  +lastTaskTypeId: ?string,
  +organization: string,
};

export type APITimeInterval = {
  paymentInterval: {
    month: number,
    year: number,
  },
  durationInSeconds: number,
};
export type APIUserLoggedTime = {
  loggedTime: Array<APITimeInterval>,
};

export type APIActiveUser = {
  email: string,
  firstName: string,
  lastName: string,
  activeTasks: number,
};

export type APIRestrictions = {|
  +allowAccess: boolean,
  +allowUpdate: boolean,
  +allowFinish: boolean,
  +allowDownload: boolean,
  +allowSave?: boolean,
|};

export type APIAllowedMode = "orthogonal" | "oblique" | "flight" | "volume";

export type APISettings = {|
  +allowedModes: Array<APIAllowedMode>,
  +preferredMode?: APIAllowedMode,
  +branchPointsAllowed: boolean,
  +somaClickingAllowed: boolean,
  +mergerMode?: boolean,
  +allowedMagnifications?: {
    shouldRestrict: boolean,
    min?: number,
    max?: number,
  },
|};

export const APIAnnotationTypeEnum = Enum.make({
  Explorational: "Explorational",
  Task: "Task",
  View: "View",
  CompoundTask: "CompoundTask",
  CompoundProject: "CompoundProject",
  CompoundTaskType: "CompoundTaskType",
});

export type APIAnnotationType = $Keys<typeof APIAnnotationTypeEnum>;

export type TracingType = "skeleton" | "volume" | "hybrid";

export type APITaskType = {
  +id: string,
  +summary: string,
  +description: string,
  +teamId: string,
  +teamName: string,
  +settings: APISettings,
  +recommendedConfiguration: ?RecommendedConfiguration,
  +tracingType: TracingType,
};

export type TaskStatus = { +open: number, +active: number, +finished: number };

type APIScriptTypeBase = {
  +name: string,
  +gist: string,
};

export type APIScript = APIScriptTypeBase & {
  +id: string,
  +owner: APIUserBase,
};

export type APIScriptUpdater = APIScriptTypeBase & {
  +id: string,
  +owner: string,
};

export type APIScriptCreator = APIScriptTypeBase & {
  +owner: string,
};

type APIProjectTypeBase = {
  +name: string,
  +team: string,
  +priority: number,
  +paused: boolean,
  +expectedTime: number, // Also known as "time limit"
  +isBlacklistedFromReport: boolean,
};

export type APIProject = APIProjectTypeBase & {
  +id: string,
  +owner: APIUserBase,
};

export type APIProjectUpdater = APIProjectTypeBase & {
  +id: string,
  +owner: string,
};

export type APIProjectCreator = APIProjectTypeBase & {
  +owner: string,
};

export type APIProjectWithAssignments = APIProject & {
  +numberOfOpenAssignments: number,
};

export type APITask = {
  +boundingBox: ?BoundingBoxObject,
  +boundingBoxVec6?: Vector6,
  +created: number,
  +creationInfo: ?string,
  +dataSet: string,
  +editPosition: Vector3,
  +editRotation: Vector3,
  +formattedHash: string,
  +id: string,
  +neededExperience: {
    +domain: string,
    +value: number,
  },
  +projectName: string,
  +script: ?APIScript,
  +status: TaskStatus,
  +team: string,
  +tracingTime: ?number,
  +type: APITaskType,
  +directLinks?: Array<string>,
};

export type APIAnnotationCompact = {
  +tracing: {
    +skeleton: ?string,
    +volume: ?string,
  },
  +dataSetName: string,
  +organization: string,
  +description: string,
  +formattedHash: string,
  +modified: number,
  +id: string,
  +isPublic: boolean,
  +name: string,
  +state: string,
  +stats: SkeletonTracingStats | {||},
  +tags: Array<string>,
  +tracingTime: ?number,
  +typ: APIAnnotationType,
};

export type LocalMeshMetaData = {|
  isVisible?: boolean,
  isLoaded?: boolean,
  isLoading?: boolean,
|};

export type RemoteMeshMetaData = {|
  annotationId: string,
  position: Vector3,
  description: string,
  id: string,
|};

export type MeshMetaData = {|
  ...LocalMeshMetaData,
  ...RemoteMeshMetaData,
|};

type APIAnnotationBase = APIAnnotationCompact & {
  +dataStore: APIDataStore,
  +tracingStore: APITracingStore,
  +restrictions: APIRestrictions,
  +settings: APISettings,
  +user?: APIUserBase,
  +meshes: Array<MeshMetaData>,
};

export type APIAnnotation = APIAnnotationBase & {
  +task: ?APITask,
};

export type APIAnnotationWithTask = APIAnnotationBase & {
  +task: APITask,
};

export type APITaskWithAnnotation = APITask & {
  +annotation: APIAnnotation,
};

export type DatasetConfig = {
  +name: string,
  +organization: string,
  +datastore: string,
  +zipFile: File,
};

type NeuroglancerLayer = {
  // This is the source URL of the layer, should start with gs://, http:// or https://
  source: string,
  type: "image" | "segmentation",
};

type NeuroglancerDatasetConfig = {
  [organizationName: string]: {
    [datasetName: string]: {
      layers: { [layerName: string]: NeuroglancerLayer },
      credentials?: Object,
    },
  },
};

type BossDatasetConfig = {
  [organizationName: string]: {
    [datasetName: string]: {
      domain: string,
      collection: string,
      experiment: string,
      username: string,
      password: string,
    },
  },
};

export type WkConnectDatasetConfig = {
  neuroglancer?: NeuroglancerDatasetConfig,
  boss?: BossDatasetConfig,
};

export type APITimeTracking = {
  time: string,
  timestamp: number,
  annotation: string,
  _id: string,
  task_id: string,
  project_name: string,
  tasktype_id: string,
  tasktype_summary: string,
};

export type APIProjectProgressReport = {
  +projectName: string,
  +paused: boolean,
  +totalTasks: number,
  +totalInstances: number,
  +openInstances: number,
  +activeInstances: number,
  +finishedInstances: number,
  +priority: number,
  +billedMilliseconds: number,
};

export type APIOpenTasksReport = {
  +id: string,
  +user: string,
  +totalAssignments: number,
  +assignmentsByProjects: { [projectName: string]: number },
};

export type APIOrganization = {
  +id: string,
  +name: string,
  +additionalInformation: string,
  +displayName: string,
  +enableAutoVerify: boolean,
};

export type APIBuildInfo = {
  webknossos: {
    name: string,
    commitHash: string,
    scalaVersion: string,
    version: string,
    sbtVersion: string,
    commitDate: string,
    ciTag: string,
    ciBuild: string,
    gitTag: string,
    version: string,
    datastoreApiVersion: string,
  },
  "webknossos-wrap": {
    builtAtMillis: string,
    name: string,
    commitHash: string,
    scalaVersion: string,
    version: string,
    sbtVersion: string,
    builtAtString: string,
  },
  webknossosDatastore?: {
    name: string,
    commitHash: string,
    scalaVersion: string,
    version: string,
    sbtVersion: string,
    commitDate: string,
    ciTag: string,
    ciBuild: string,
    gitTag: string,
    version: string,
    datastoreApiVersion: string,
  },
};

export type APIFeatureToggles = {
  +discussionBoard: string | false,
  +discussionBoardRequiresAdmin: boolean,
  +defaultOrganization: string,
  +addForeignDataset: boolean,
  +hideNavbarLogin: boolean,
  +isDemoInstance: boolean,
  +autoBrushReadyDatasets: Array<string>,
  +isDemoInstance: boolean,
};

// Tracing related datatypes
export type APIUpdateActionBatch = {
  version: number,
  value: Array<ServerUpdateAction>,
};

export type ServerNode = {
  id: number,
  position: Point3,
  rotation: Point3,
  bitDepth: number,
  viewport: number,
  resolution: number,
  radius: number,
  createdTimestamp: number,
  interpolation: boolean,
};

export type ServerBranchPoint = {
  createdTimestamp: number,
  nodeId: number,
};

export type ServerBoundingBox = {
  topLeft: Point3,
  width: number,
  height: number,
  depth: number,
};

export type ServerBoundingBoxTypeTuple = {
  topLeft: Vector3,
  width: number,
  height: number,
  depth: number,
};

export type ServerSkeletonTracingTree = {
  branchPoints: Array<ServerBranchPoint>,
  color: ?{ r: number, g: number, b: number },
  comments: Array<CommentType>,
  edges: Array<Edge>,
  name: string,
  nodes: Array<ServerNode>,
  treeId: number,
  createdTimestamp: number,
  groupId?: ?number,
  isVisible?: boolean,
};

export type ServerTracingBase = {|
  id: string,
  userBoundingBox?: ServerBoundingBox,
  createdTimestamp: number,
  dataSetName: string,
  editPosition: Point3,
  editRotation: Point3,
  error?: string,
  version: number,
  zoomLevel: number,
|};

export type ServerSkeletonTracing = {|
  ...ServerTracingBase,
  activeNodeId?: number,
  boundingBox?: ServerBoundingBox,
  trees: Array<ServerSkeletonTracingTree>,
  treeGroups: ?Array<TreeGroup>,
|};

export type ServerVolumeTracing = {|
  ...ServerTracingBase,
  activeSegmentId?: number,
  boundingBox: ServerBoundingBox,
  elementClass: ElementClass,
  fallbackLayer?: string,
  largestSegmentId: number,
|};

export type ServerTracing = ServerSkeletonTracing | ServerVolumeTracing;

export type HybridServerTracing = {
  skeleton: ?ServerSkeletonTracing,
  volume: ?ServerVolumeTracing,
};

export default {};
