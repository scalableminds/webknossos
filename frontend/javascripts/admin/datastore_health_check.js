// @flow
import _ from "lodash";

import Request from "libs/request";
import * as RestAPI from "admin/admin_rest_api";
import Toast from "libs/toast";
import messages from "messages";

// Create a throttled function which depends on its arguments.
// That way, each datastore is checked for health in a throttled and isolated manner
const memoizedThrottle = (func, wait = 0, options = {}): Function => {
  // Memoize the creation of a throttling function
  const mem = _.memoize(() => _.throttle(func, wait, options), options.resolver);
  return (...args: Array<*>) => {
    // look up (or create) the throttling function and invoke it
    mem(...args)(...args);
  };
};

// Do not call this function directly, but call pingMentionedDataStores instead
// which will take care of extracting the hostnames.
// Otherwise, the memoization will not work correctly if the path and query-string are part of the URL
const pingDataStoreIfAppropriate = memoizedThrottle(async (requestedUrl: string): Promise<*> => {
  const [datastores, tracingstore, isInMaintenance] = await Promise.all([
    RestAPI.getDataStoresCached(),
    RestAPI.getTracingStoreCached(),
    RestAPI.isInMaintenance(),
  ]).catch(() => [null, null, null]);
  if (datastores == null || tracingstore == null || isInMaintenance == null) {
    Toast.warning(messages.offline);
    return;
  }

  const stores = datastores
    .map(datastore => ({ ...datastore, path: "data" }))
    .concat({ ...tracingstore, path: "tracings" });
  if (isInMaintenance) {
    Toast.warning(messages.planned_maintenance);
  } else {
    const usedStore = stores.find(ds => requestedUrl.indexOf(ds.url) > -1);
    if (usedStore != null) {
      const { url, path } = usedStore;
      const healthEndpoint = `${url}/${path}/health`;
      Request.triggerRequest(healthEndpoint, {
        doNotInvestigate: true,
        mode: "cors",
        timeout: 5000,
      }).then(
        () => checkVersionMismatch(url),
        () => Toast.warning(messages["datastore.health"]({ url })),
      );
    }
  }
}, 5000);

async function checkVersionMismatch(url: string) {
  const [buildinfoWebknossos, buildinfoDatastore] = await Promise.all([
    RestAPI.getBuildInfo(),
    RestAPI.getDataStoreBuildInfo(url),
  ]);

  const expectedDatastoreApiVersion = buildinfoWebknossos.webknossos.datastoreApiVersion;

  const buildInfoWebknossosDatastore = buildinfoDatastore.webknossosDatastore
    ? buildinfoDatastore.webknossosDatastore
    : buildinfoDatastore.webknossos;
  const suppliedDatastoreApiVersion = buildInfoWebknossosDatastore.datastoreApiVersion;

  if (
    Number(expectedDatastoreApiVersion.split(".")[0]) <
    Number(suppliedDatastoreApiVersion.split(".")[0])
  ) {
    Toast.warning(
      messages["datastore.version.too_new"]({
        expectedDatastoreApiVersion,
        suppliedDatastoreApiVersion,
        url,
      }),
    );
  } else if (
    Number(expectedDatastoreApiVersion.split(".")[0]) >
    Number(suppliedDatastoreApiVersion.split(".")[0])
  ) {
    Toast.warning(
      messages["datastore.version.too_old"]({
        expectedDatastoreApiVersion,
        suppliedDatastoreApiVersion,
        url,
      }),
    );
  }
}

const extractUrls = (str: string): Array<string> => {
  const urlMatcher = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{2,256}(\.[a-z]{2,6})?\b([-a-zA-Z0-9@:%_+.~#?&\\=]*)/g;
  return _.uniq(str.match(urlMatcher) || []);
};

export const pingMentionedDataStores = (str: string): void => {
  extractUrls(str).map(pingDataStoreIfAppropriate);
};

export default {};
