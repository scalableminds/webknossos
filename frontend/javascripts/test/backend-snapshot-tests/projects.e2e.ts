import _ from "lodash";
import type { APIProject, APIProjectUpdater } from "types/api_flow_types";
import {
  tokenUserA,
  tokenUserD,
  setCurrToken,
  resetDatabase,
  writeTypeCheckingFile,
} from "test/enzyme/e2e-setup";
import * as api from "admin/admin_rest_api";
import test from "ava";
test.before("Reset database", async () => {
  resetDatabase();
});
test.beforeEach("Change token", async () => {
  setCurrToken(tokenUserA);
});
test.serial("getProjects()", async (t) => {
  const projects = _.sortBy(await api.getProjects(), (p) => p.name);

  writeTypeCheckingFile(projects, "project", "APIProject", {
    isArray: true,
  });
  t.snapshot(projects, {
    id: "projects-getProjects()",
  });
});
test.serial("getProjectsWithOpenAssignments()", async (t) => {
  const projects = _.sortBy(await api.getProjectsWithOpenAssignments(), (p) => p.name);

  t.snapshot(projects, {
    id: "projects-getProjectsWithOpenAssignments()",
  });
});
test.serial("getProject(projectId: string)", async (t) => {
  const projectId = _.sortBy(await api.getProjects(), (p) => p.name)[0].id;

  const project = await api.getProject(projectId);
  t.snapshot(project, {
    id: "projects-getProject(projectId: string)",
  });
});
test.serial("createProject and deleteProject", async (t) => {
  const team = _.sortBy(await api.getTeams(), (aTeam) => aTeam.name)[0];

  const activeUser = await api.getActiveUser();
  const projectName = "test-new-project";
  const newProject = {
    name: projectName,
    team: team.id,
    teamName: team.name,
    owner: activeUser.id,
    priority: 1,
    paused: false,
    expectedTime: 1,
    numberOfOpenAssignments: 1,
    isBlacklistedFromReport: true,
  };
  const createdProject = await api.createProject(newProject);
  // Since the id will change after re-runs, we fix it here for easy
  // snapshotting
  const createdProjectWithFixedId = Object.assign({}, createdProject, {
    id: "fixed-project-id",
  });
  t.snapshot(createdProjectWithFixedId, {
    id: "projects-createProject(project: APIProjectType)",
  });
  const response = await api.deleteProject(createdProject.id);
  t.snapshot(response, {
    id: "projects-deleteProject(projectId: string)",
  });
});

function convertProjectToProjectUpdater(project: APIProject): APIProjectUpdater {
  return Object.assign({}, project, {
    owner: project.owner.id,
  });
}

test.serial("updateProject(projectId: string, project: APIProjectType)", async (t) => {
  const project = (await api.getProjects())[0];
  const projectId = project.id;
  const projectWithOwnerId = convertProjectToProjectUpdater(project);
  const projectWithNewPriority: APIProjectUpdater = Object.assign({}, projectWithOwnerId, {
    priority: 1337,
  });
  const updatedProject = await api.updateProject(projectId, projectWithNewPriority);
  t.snapshot(updatedProject, {
    id: "projects-updateProject(projectId: string, project: APIProjectType)",
  });
  const revertedProject = await api.updateProject(projectId, projectWithOwnerId);
  t.snapshot(revertedProject, {
    id: "projects-revertedProject",
  });
});
test.serial("increaseProjectTaskInstances", async (t) => {
  await setCurrToken(tokenUserD);
  const projectId = (await api.getProjects())[0].id;
  const updatedProject = await api.increaseProjectTaskInstances(projectId, 10);
  t.snapshot(updatedProject, {
    id: "projects-increaseProjectTaskInstances(projectId: string, delta?: number)",
  });
});
test.serial("pauseProject and resumeProject", async (t) => {
  const projectId = (await api.getProjects())[0].id;
  const pausedProject = await api.pauseProject(projectId);
  t.snapshot(pausedProject, {
    id: "projects-pauseProject(projectId: string)",
  });
  const resumedProject = await api.resumeProject(projectId);
  t.snapshot(resumedProject, {
    id: "projects-resumeProject(projectId: string)",
  });
});
