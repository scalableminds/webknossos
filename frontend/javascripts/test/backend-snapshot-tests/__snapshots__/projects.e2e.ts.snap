// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Project API (E2E) > createProject and deleteProject 1`] = `
{
  "created": "created",
  "expectedTime": 60000,
  "id": "id",
  "isBlacklistedFromReport": true,
  "name": "test-new-project",
  "owner": {
    "email": "user_A@scalableminds.com",
    "firstName": "user_A",
    "id": "id",
    "isAdmin": true,
    "isAnonymous": false,
    "isDatasetManager": true,
    "lastName": "last_A",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X1",
      },
      {
        "id": "id",
        "isTeamManager": false,
        "name": "team_X3",
      },
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X4",
      },
    ],
  },
  "paused": false,
  "priority": 1,
  "team": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
}
`;

exports[`Project API (E2E) > createProject and deleteProject 2`] = `
{
  "messages": [
    {
      "success": "Project was removed successfully",
    },
  ],
}
`;

exports[`Project API (E2E) > getProject(projectId: string) 1`] = `
{
  "created": "created",
  "expectedTime": 90,
  "id": "id",
  "isBlacklistedFromReport": false,
  "name": "Test_Project",
  "owner": {
    "email": "user_A@scalableminds.com",
    "firstName": "user_A",
    "id": "id",
    "isAdmin": true,
    "isAnonymous": false,
    "isDatasetManager": true,
    "lastName": "last_A",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X1",
      },
      {
        "id": "id",
        "isTeamManager": false,
        "name": "team_X3",
      },
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X4",
      },
    ],
  },
  "paused": false,
  "priority": 100,
  "team": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
}
`;

exports[`Project API (E2E) > getProjects() 1`] = `
[
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project",
    "owner": {
      "email": "user_A@scalableminds.com",
      "firstName": "user_A",
      "id": "id",
      "isAdmin": true,
      "isAnonymous": false,
      "isDatasetManager": true,
      "lastName": "last_A",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X1",
        },
        {
          "id": "id",
          "isTeamManager": false,
          "name": "team_X3",
        },
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X4",
        },
      ],
    },
    "paused": false,
    "priority": 100,
    "team": "570b9f4b2a7c0e3b008da6ec",
    "teamName": "team_X1",
  },
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project2",
    "owner": {
      "email": "user_D@scalableminds.com",
      "firstName": "user_D",
      "id": "id",
      "isAdmin": false,
      "isAnonymous": false,
      "isDatasetManager": false,
      "lastName": "last_D",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X2",
        },
      ],
    },
    "paused": false,
    "priority": 90,
    "team": "69882b370d889b84020efd4f",
    "teamName": "team_X2",
  },
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project3(for_annotation_mutations)",
    "owner": {
      "email": "user_D@scalableminds.com",
      "firstName": "user_D",
      "id": "id",
      "isAdmin": false,
      "isAnonymous": false,
      "isDatasetManager": false,
      "lastName": "last_D",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X2",
        },
      ],
    },
    "paused": false,
    "priority": 100,
    "team": "69882b370d889b84020efd4f",
    "teamName": "team_X2",
  },
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project4",
    "owner": {
      "email": "user_A@scalableminds.com",
      "firstName": "user_A",
      "id": "id",
      "isAdmin": true,
      "isAnonymous": false,
      "isDatasetManager": true,
      "lastName": "last_A",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X1",
        },
        {
          "id": "id",
          "isTeamManager": false,
          "name": "team_X3",
        },
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X4",
        },
      ],
    },
    "paused": false,
    "priority": 200,
    "team": "570b9f4b2a7c0e3b008da6ec",
    "teamName": "team_X1",
  },
]
`;

exports[`Project API (E2E) > getProjectsWithStatus() 1`] = `
[
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project",
    "owner": {
      "email": "user_A@scalableminds.com",
      "firstName": "user_A",
      "id": "id",
      "isAdmin": true,
      "isAnonymous": false,
      "isDatasetManager": true,
      "lastName": "last_A",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X1",
        },
        {
          "id": "id",
          "isTeamManager": false,
          "name": "team_X3",
        },
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X4",
        },
      ],
    },
    "paused": false,
    "pendingInstances": 19,
    "priority": 100,
    "team": "570b9f4b2a7c0e3b008da6ec",
    "teamName": "team_X1",
    "tracingTime": "tracingTime",
  },
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project2",
    "owner": {
      "email": "user_D@scalableminds.com",
      "firstName": "user_D",
      "id": "id",
      "isAdmin": false,
      "isAnonymous": false,
      "isDatasetManager": false,
      "lastName": "last_D",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X2",
        },
      ],
    },
    "paused": false,
    "pendingInstances": 10,
    "priority": 90,
    "team": "69882b370d889b84020efd4f",
    "teamName": "team_X2",
    "tracingTime": "tracingTime",
  },
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project3(for_annotation_mutations)",
    "owner": {
      "email": "user_D@scalableminds.com",
      "firstName": "user_D",
      "id": "id",
      "isAdmin": false,
      "isAnonymous": false,
      "isDatasetManager": false,
      "lastName": "last_D",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X2",
        },
      ],
    },
    "paused": false,
    "pendingInstances": 10,
    "priority": 100,
    "team": "69882b370d889b84020efd4f",
    "teamName": "team_X2",
    "tracingTime": "tracingTime",
  },
  {
    "created": "created",
    "expectedTime": 90,
    "id": "id",
    "isBlacklistedFromReport": false,
    "name": "Test_Project4",
    "owner": {
      "email": "user_A@scalableminds.com",
      "firstName": "user_A",
      "id": "id",
      "isAdmin": true,
      "isAnonymous": false,
      "isDatasetManager": true,
      "lastName": "last_A",
      "teams": [
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X1",
        },
        {
          "id": "id",
          "isTeamManager": false,
          "name": "team_X3",
        },
        {
          "id": "id",
          "isTeamManager": true,
          "name": "team_X4",
        },
      ],
    },
    "paused": false,
    "pendingInstances": 0,
    "priority": 200,
    "team": "570b9f4b2a7c0e3b008da6ec",
    "teamName": "team_X1",
    "tracingTime": "tracingTime",
  },
]
`;

exports[`Project API (E2E) > increaseProjectTaskInstances 1`] = `
{
  "created": "created",
  "expectedTime": 90,
  "id": "id",
  "isBlacklistedFromReport": false,
  "name": "Test_Project2",
  "owner": {
    "email": "user_D@scalableminds.com",
    "firstName": "user_D",
    "id": "id",
    "isAdmin": false,
    "isAnonymous": false,
    "isDatasetManager": false,
    "lastName": "last_D",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X2",
      },
    ],
  },
  "paused": false,
  "pendingInstances": 20,
  "priority": 90,
  "team": "69882b370d889b84020efd4f",
  "teamName": "team_X2",
  "tracingTime": "tracingTime",
}
`;

exports[`Project API (E2E) > pauseProject and resumeProject 1`] = `
{
  "created": "created",
  "expectedTime": 90,
  "id": "id",
  "isBlacklistedFromReport": false,
  "name": "Test_Project",
  "owner": {
    "email": "user_A@scalableminds.com",
    "firstName": "user_A",
    "id": "id",
    "isAdmin": true,
    "isAnonymous": false,
    "isDatasetManager": true,
    "lastName": "last_A",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X1",
      },
      {
        "id": "id",
        "isTeamManager": false,
        "name": "team_X3",
      },
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X4",
      },
    ],
  },
  "paused": true,
  "priority": 100,
  "team": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
}
`;

exports[`Project API (E2E) > pauseProject and resumeProject 2`] = `
{
  "created": "created",
  "expectedTime": 90,
  "id": "id",
  "isBlacklistedFromReport": false,
  "name": "Test_Project",
  "owner": {
    "email": "user_A@scalableminds.com",
    "firstName": "user_A",
    "id": "id",
    "isAdmin": true,
    "isAnonymous": false,
    "isDatasetManager": true,
    "lastName": "last_A",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X1",
      },
      {
        "id": "id",
        "isTeamManager": false,
        "name": "team_X3",
      },
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X4",
      },
    ],
  },
  "paused": false,
  "priority": 100,
  "team": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
}
`;

exports[`Project API (E2E) > updateProject(projectId: string, project: APIProjectType) 1`] = `
{
  "created": "created",
  "expectedTime": 5400000,
  "id": "id",
  "isBlacklistedFromReport": false,
  "name": "Test_Project",
  "owner": {
    "email": "user_A@scalableminds.com",
    "firstName": "user_A",
    "id": "id",
    "isAdmin": true,
    "isAnonymous": false,
    "isDatasetManager": true,
    "lastName": "last_A",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X1",
      },
      {
        "id": "id",
        "isTeamManager": false,
        "name": "team_X3",
      },
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X4",
      },
    ],
  },
  "paused": false,
  "priority": 1337,
  "team": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
}
`;

exports[`Project API (E2E) > updateProject(projectId: string, project: APIProjectType) 2`] = `
{
  "created": "created",
  "expectedTime": 5400000,
  "id": "id",
  "isBlacklistedFromReport": false,
  "name": "Test_Project",
  "owner": {
    "email": "user_A@scalableminds.com",
    "firstName": "user_A",
    "id": "id",
    "isAdmin": true,
    "isAnonymous": false,
    "isDatasetManager": true,
    "lastName": "last_A",
    "teams": [
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X1",
      },
      {
        "id": "id",
        "isTeamManager": false,
        "name": "team_X3",
      },
      {
        "id": "id",
        "isTeamManager": true,
        "name": "team_X4",
      },
    ],
  },
  "paused": false,
  "priority": 100,
  "team": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
}
`;
